### PR TITLE
Refactor password checker, add extender

### PR DESCRIPTION
**Pull request by askvortsov1**
**Part of #1891**

**Changes proposed in this pull request:**
- Deprecate CheckingPassword event
- Use callbacks for registering password checkers
- Add extender for adding/removing password checkers
- Add integration tests for password checkers

**Reviewers should focus on:**
- Anything I missed?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).

### CODE REVIEW

File: src/User/User.php
Comment (luceos) [inline src/User/User.php:353]: What's the benefit of not immediately returning false here. In any future iteration the value will only be set to true if true anyway. Allowing more iteration will only cost us performance here.

File: src/User/UserServiceProvider.php
Comment (luceos) [inline src/User/UserServiceProvider.php:108]: Newline isn't really necessary here?

File: src/User/UserServiceProvider.php
Comment (luceos) [inline src/User/UserServiceProvider.php:105]: I've seen similar logic before, can't we make a helper function that loops and resolves automatically?

Comment (askvortsov1) [reply src/User/UserServiceProvider.php:105]: We could, but considering the need for imports, I don't know how much better it would be. Also unsure where we would put it. Regardless, I'd rather revisit this after we finish extenders.

File: src/User/User.php
Comment (askvortsov1) [reply src/User/User.php:353]: Do you mean returning true? It's because an output of false overrides any outputs of true, regardless of order of evaluation. Unless order is a feature (as with middleware), I'm trying to maximally avoid order of evaluation mattering. 

As such, an explicit false, denies, an explicit true allows as long as there are no explicit denials, and no trues means deny.

File: src/Extend/User.php
Comment (franzliedke) [inline src/Extend/User.php:44]: Is there a known use-case for this currently? If not, I'd avoid it for now...

**Comment by franzliedke** [comment]
On a higher level: This feels a bit too oriented around the old event, but I'd like us to focus on *use cases* with the extenders. What are we (extension authors) hoping to achieve when using this extender?

Doesn't changing the password checker usually mean we'd need to change the hasher as well? We might need an abstraction around this combination of actions: persisting and checking passwords. (PHP's password hashing abstractions also adds a rehash check to the mix.)

**Comment by askvortsov1** [comment]
Well, I'd agree with that point. This reminds me most of Django's authentication backends: https://docs.djangoproject.com/en/3.0/topics/auth/customizing/. We could expand the scope of the extender to entail that?

**Comment by askvortsov1** [comment]
Shoo!

**Comment by askvortsov1** [comment]
@clarkwinkelmann The only examples of real-life use of this event are in your extensions, so congrats, you're hereby subpoenaed as an expert witness :stuck_out_tongue:!

I agree with Franz that the password checker and hasher are related concepts. But from what I'm seeing in https://github.com/clarkwinkelmann/flarum-ext-passwordless/blob/230ef8a5602bb4a72a25a42f50823d142d44f932/src/Listeners/CheckPassword.php and https://github.com/migratetoflarum/old-passwords/blob/d6b78d2ac6cd609d41e5df1af69c548bcda920f7/src/Listeners/CheckPassword.php, it's not _quite_ that direct. IMO, we have a few options:

- Add methods to the extender to change the password hasher, this might be a good idea regardless (but what if multiple extensions do it at once?)
- Combine the hasher and password checker into a "password backend" driver. Not sure I like this idea though
- Go a step beyond, and substitute the whole "checking password" method with a Django-esque auth driver. But honestly that's pretty similar to what we have now.

File: src/User/User.php
Comment (clarkwinkelmann) [reply src/User/User.php:362]: So if I understand correctly, we need at least one return to be `=== true` and none `=== false` for `checkPassword` to return `true`.

I'm assuming it's so checks can return `null` without impacting the result if they have neither success nor failure.

File: src/Extend/Auth.php
Comment (clarkwinkelmann) [inline src/Extend/Auth.php:24]: Needs `callable|string` docblock

File: src/User/UserServiceProvider.php
Comment (clarkwinkelmann) [inline src/User/UserServiceProvider.php:103]: I'm concerned that this gets resolved every time Flarum boots, even though password validation is something that happens very rarely during the lifetime of the app. I'd advocate a solution similar to https://github.com/flarum/core/pull/2423#discussion_r517028321

**Review by clarkwinkelmann — COMMENTED** [review]
Indeed, seems like I'm the only one to have attempted to extend this.

One use case is definitely replacing the default hasher. But this is probably a rare use case, since the default is very sensible.

On the hasher front, I'd suggest the following, if that's not already the case: allow customizing the hasher of the Laravel Hash object with a config value (assuming this is customizable with a constructor parameter), and have the hasher in the container (that's already the case).

This offers a lot of flexibility and an extension can always replace the hasher in the container if they wish so. No need to build any extender or configuration UI for the password hasher IMO.

---

The other use case is the one I have used in these two occasions: allowing additional things to be used as password, which aren't in the `password` column, and possibly not in the database at all.

I think the proposed solution in this PR looks good. I don't think we really *need* priority to be customizable. Ability to add and remove should be enough. Having the default check be in the stack also allows extensions to use this path if they want to replace the default hasher with something that's not plug-and-play into the Laravel Hasher.

File: src/Extend/Auth.php
Comment (askvortsov1) [reply src/Extend/Auth.php:24]: For the hasher, couldn't an extension just override `$container->make('hash')` via a ServiceProvider extender?

File: src/User/UserServiceProvider.php
Comment (askvortsov1) [reply src/User/UserServiceProvider.php:103]: I'll update this when #2423 is merged

**Review by clarkwinkelmann — COMMENTED** [review]
Any reason `ContainerUtil::wrapCallback` is called in the provider and not in the extender? Are we consistent across extenders with that way of doing things?

**Comment by clarkwinkelmann** [comment]
@askvortsov1 was something holding this back?

Should I do a final review now or wait for something else?

**Comment by askvortsov1** [comment]
I don't believe so, should be good to go

**Comment by askvortsov1** [comment]
Merge conflicts resolved, this should be good to review.

**Review by clarkwinkelmann — CHANGES_REQUESTED** [review]
I have not tested locally, but it's looking all good to me.

I'll try to update my old-passwords extension before release to actually test it with the new system once this is merged to master.

I think you will want to update the deprecation comment before merging? It's the only thing I noticed.

File: src/User/Event/CheckingPassword.php
Comment (clarkwinkelmann) [reply src/User/Event/CheckingPassword.php:15]: Also, I think the name of the extender in the comment is wrong?

**Review by clarkwinkelmann — APPROVED** [review]

**Review by tankerkiller125 — APPROVED** [review]

File: src/User/User.php
Comment (askvortsov1) [reply src/User/User.php:354]: To clarify, checkers should return as follows:

- If the password is valid, return true
- If not, return null
- If the request must be blocked, return false

This is what's currently used in Clark's extensions. Before I merge this in, is this schema fine with everyone?

Comment (clarkwinkelmann) [reply src/User/User.php:354]: Maybe we could offer CONST aliases just like we did for policies? I'm ok with the raw value, but maybe it would be a bit easier in the documentation if each of these option had a human readable constant associated with it?

Comment (askvortsov1) [reply src/User/User.php:354]: Hmm, I'm not sure. They'd have to be properties of the User object. More importantly though, I have no idea what to call the `null` one. `User::PASSWORD_VALID` and `User::PASSWORD_INVALID` make sense, but maybe `User::PASSWORD_SKIP`?

Comment (clarkwinkelmann) [reply src/User/User.php:354]: Not sure :upside_down_face: 

Maybe I'd call them something including AUTHORIZE, FAIL_AUTHORIZATION and something with PASS/SKIP/NEXT?

Comment (clarkwinkelmann) [reply src/User/User.php:354]: Also, technically you don't really need to return `null` ever. It's also what happens when you don't return anything so there might be a need for a constant.

Comment (askvortsov1) [reply src/User/User.php:354]: I think that this might make things more confusing to be completely honest. Let's keep it as is for now; if we see confusion in the developer community we can always add aliases later I think?

> Also, technically you don't really need to return null ever. It's also what happens when you don't return anything so there might not be a need for a constant.

True, I'll try to clarify this in the docblock

**Pull request merged**
